### DISCUSSION TITLE
New Tree Support: Coordinate Remaining Tasks

### DISCUSSION THREAD

**Discussion by rburema** [Ultimaker/Cura#14805] [Collaboration]
This set of discussions is meant to coordinate work between developers both in and external to UltiMaker, w.r.t. the New Tree Support, specifically as it relates to the release of the first Cura version that it's going to be in.

More precisely, we will post a few sub-discussions with the tasks that we still think we need to do. (From the UM side this will probably mean Jira tickets).

---

For a more end-user focussed discussion, and/or reporting new problems, please use either the normal Github bug-reporting flow, and/or this discussion about the alpha we already released with a version of the new tree support.

---

'Tree Support 2.0', or 'New Tree Support' represents a significant improvement to (tree) support, and we _plan_ to replace the current version of Tree Support with this in 5.4.0., _if_ all goes well.

(You can already try this out: There is an early alpha, based from 5.2.x and mildly confusingly called [Cura 5.3.0-Alpha-Xmas](https://ultimaker.com/learn/cura-christmas-tree-support).)

For the original Pull Request(s) related to this see [here for the engine one](https://github.com/Ultimaker/CuraEngine/pull/1282). For settings added or changed in the front-end see [this branch](https://github.com/Ultimaker/Cura/tree/5.3.0-alpha%2Bxmas).

**Comment by rburema** [comment]
CURA-10062 Missing Support Interface

**Reply by rburema** [reply]
Description:

When Interface priority is set to preferred, there is missing interface for some branches. 

The interface should follow the model, not the support.

**Reply by ThomasRahm** [reply]
There was a related discussion about this topic in the [pull request](https://github.com/Ultimaker/CuraEngine/pull/1282) (Messages from 29.12 to 4.1. should be about this)

I need to check if [my point](https://github.com/Ultimaker/CuraEngine/pull/1282#issuecomment-1369876316) from back then still stands with the support changes in 5.3. I will try to get my development instance to Cura 5.3 within the week.

**Reply by rburema** [reply]
Hi! Sorry, I was just setting these up & then things came in-between.

You're indeed already aware of this one.

Some more info from the ticket (in the analysis section):

Proposed solutions:

Redo support interface areas for tree support to match the behaviour of support interface areas for normal support. [More details](https://github.com/Ultimaker/CuraEngine/pull/1282#issuecomment-1369876316)

Tune horizontal expansion settings:

![image-20230104-104213](https://user-images.githubusercontent.com/41987080/223858010-9cfd8c5c-3f4f-4306-9dbd-67c6da4ccb49.png)


Preferably you want the most right situation without the extra support generated. The extra support is generated because the support folds around the model. This is the same issue we currently have with normal support. The idea of [CURA-10119: Spijk with Rijk: avoid support folding around model due to support horizontal expansionDONE](https://ultimaker.atlassian.net/browse/CURA-10119) is to come up with a solution for this problem.

**Reply by ThomasRahm** [reply]
I am currently looking into this one as i have an idea for it.
I still have some edge-cases to fix, but with `Support Roof Walls` set to 1 the results look promising.

![grafik](https://user-images.githubusercontent.com/67757218/224516100-dee0e94e-c0b4-49a7-92d3-d30e90d61ade.png)

![grafik](https://user-images.githubusercontent.com/67757218/224516120-f20ab053-bf6f-4ca3-8187-7ca188a00ad3.png)


I do however have a the replacing tips with roof hard-coded to false (in the screenshots), also i have `Minimum Support Roof Area` set to 0, which is not optimal.
So (hopefully) aside from fixing a few edge-cases i just need to figure out when to apply which behavior based on the settings.

**Reply by ThomasRahm** [reply]
I encountered some edge-cases that require me to do the roof generation differently, and while doing this i got fed up by the size of the generateInitialAreas function, so i am refactoring it to be its own class. 

I also noticed an issue with my old implementation (roof is always a bit (should be about tip radius) too far from the model) while working on the refactor.

**Reply by ThomasRahm** [reply]
I think i am mostly done, and pushed my changes. I may need to do some clean-up, which i will do in another commit. Not all changes as seen by git are actually changes as quite a few functions migrated from the TreeSupport class to the new TreeSupportTipGenerator and TreeSupportUtils classes.
What tool are you using to format the code? My IDE seems to go for another style, especially with the parallel for loops.

I did only do some testing, so it could be that i missed some problems.
My Internet is just very spotty currently and i wanted to get a version online.
To achieve the nice roof follows model effect, the support roof wall line count has to be 1 or larger.

I am a bit unsure how to determine if a overhang should be roof. Just applying the minimum support roof area seems counterproductive, as then a small area that would contribute to a larger roof area further down is removed, which in turn could cause the roof area below to fall below the minimum support roof area (and so forth).

I added another constant to switch between interpreting the minimum support roof area as a hard limit, and more of a soft limit, where i check if a roof could be part of a large enough area (and only remove it if it is not). I am looking feedback which behavior is better and why.
In the commit it is set to `SUPPORT_TREE_MINIMUM_ROOF_AREA_HARD_LIMIT = false` as i think false produces better results for small area/overhang angles, even though it may violate minimum support roof area.

**Reply by rburema** [reply]
Thanks! This does seem to be mostly working. I do see some floating interface where no trees are generated. I'll try to look at those instances myself tomorrow.

> I added another constant to switch between interpreting the minimum support roof area as a hard limit, and more of a soft limit, where i check if a roof could be part of a large enough area (and only remove it if it is not). I am looking feedback which behavior is better and why.

I'll have a think about wheter or not we (temporarily) want to make that a setting when we test things out. (At least this Friday.)

> What tool are you using to format the code? My IDE seems to go for another style, especially with the parallel for loops.

There's unfortunately no definitive answer to this presently. At the moment, we have an outdated official style guide, a linter that doesn't really fully adhere to that, and the legacy style of several developers that we never got around to changing.

The IDE's (and tools) we use are all different, a bit by design, since it keeps the code a bit more robust if we always know it can compile on different OS's with different compilers.

We should definitely finish the automatic linting ticket we have, it's just at the moment we have different priorities.

**Reply by jellespijker** [reply]
> What tool are you using to format the code? My IDE seems to go for another style, especially with the parallel for loops.

We used to be not the most consistent in our formatting, which resulted in a couple of unnecessary discussions during code reviews, which were more about form that content.

To circumvent that we added clang-format and clang-tidy files. These files are not completely according to our style (yet). But you should still use them for formatting.  If we discover during a code review that those tools don't provide the correct output we can change the configuration files in the PR according to our style. 

Since our exiating code base is inconsistent in the formatting you should take care when using them to format existing files. Apply them only on the changed sections. If you introduce new files you should apply them from the start.

This is the most pragmatic way to move towards a fully automated format and statically analysed code base.

**Reply by ThomasRahm** [reply]
> I do see some floating interface where no trees are generated.

As i now generate roof areas before i generate the tips, this is an issue that could occur more easily. I have added a function that should remove tips inside of support roof and tips that should support roof, but don't have any above it called `removeUselessAddedPoints`. This issue could also be related to it.

I got curious and looked at my favorite test model again (the hobgoblin) and while i could not see any non supported roofs, i found that a small rounding error could have caused very small roof areas to be not supported. I improved the error handling in that case a bit, but i would be surprised if this fixed your issue.
Happy to take a look if you can share the project file.

**Reply by rburema** [reply]
> Happy to take a look if you can share the project file.

Thanks!

I tried the fix you made, but no dice. I don't think this is a small rounding error. It seems like the roof is overwriting the tips in some cases, causing tree generation to fail and only leave the interface behind. (Because if I slice it without interface, it does generate.)

[UMS5_HMAL_3S.zip](https://github.com/Ultimaker/Cura/files/11037750/UMS5_HMAL_3S.zip) I've attached the model.

(In the mean time, I fixed a divide by zero error in the branch I'm working on now. https://github.com/Ultimaker/CuraEngine/tree/CURA-10415_update_tree (There's a front-end branch as well https://github.com/Ultimaker/Cura/tree/CURA-10415_update_tree )

**Reply by ThomasRahm** [reply]
Can reproduce it, found the cause and pushed a fix. Thank you :)

Forgot to ensure check that even if there is no overhang that has to be supported on a layer, the layer above may have roof that has to supported. Seems my test models have overhang on every layer.

**Reply by rburema** [reply]
Thanks again. That was even faster than you normally fix things! I'll have a look merge it in later on our side.

**Comment by rburema** [comment]
CURA-10065 Floating Islands Inside Support

**Reply by ThomasRahm** [reply]
Partially fixed since https://github.com/ThomasRahm/CuraEngine/commit/8bd759ae2fdf12abdc4c921be1782fb369bc4759
Example of a case that would be fixed:
https://github.com/Ultimaker/Cura/issues/14702

Fix can cause large overhangs in support structures can occur. See https://github.com/ThomasRahm/CuraEngine/issues/9#issuecomment-1369838446 for example and cause of the issue.

It is still an issue in some cases when interface is active.
Example: 

![image](https://user-images.githubusercontent.com/67757218/223580069-297cdb37-bb39-4e76-98a1-637f3628b81f.png)


(I was asked to not publicly share this model, so i cant upload it here)

**Reply by rburema** [reply]
I think that might indeed be it. Here's our information on it:

The issue occurs where the branches merge (where it hits the trunk). 

https://user-images.githubusercontent.com/41987080/223859800-e8055683-53da-455f-ab28-546b71f4523c.mov

Possible solution: Remove disconnected internal islands, just print the outline

**Reply by rburema** [reply]
I can confirm that this works in a lot of cases. I still see some weirdness on rare occasions. Will need to investigate whether or not it's caused by this. I'll try to see if I can figure out if it's a false alarm or not tomorrow.

**Reply by rburema** [reply]
Hm ... it does seem like your latest commit did fix this. I can't reproduce the thing I saw yesterday with the current code at least. :-)

**Reply by ThomasRahm** [reply]
It definitely should not have. All changes my commit from tonight did were related to tip placement. This would have to be related to the drawing of branches. Maybe the tips that caused this behavior moved slightly, which causes it to no longer appear with your test model.

**Reply by rburema** [reply]
Hm. Makes sense. Here is the model that went wrong for me with the old code.

(Never mind the name, it's not relevant.) 
[crash compatible printers.zip](https://github.com/Ultimaker/Cura/files/11041744/crash.compatible.printers.zip)

**Reply by ThomasRahm** [reply]
I found a huge issue with my fix for the floating lines:
If the area that simulates the support pattern is closed all (simulated) lines will be removed, as they only exist because of said holes

Example:
[test_model1_breaking_remove_hole.zip](https://github.com/Ultimaker/Cura/files/11129005/test_model1_breaking_remove_hole.zip)
In screenshots:
With getOutsidePolygons in finalizeInterfaceAndSupportAreas:
![image](https://user-images.githubusercontent.com/67757218/229279396-798171a4-0bbf-4dc4-afc4-0576ebc5c6c1.png)

Without getOutsidePolygons in finalizeInterfaceAndSupportAreas:
![image](https://user-images.githubusercontent.com/67757218/229279461-d5876c5e-2e29-4df7-b658-96a86e3352f6.png)

**Reply by ThomasRahm** [reply]
Currently looking into this. While the idea i have seems like it may work, it is too slow currently.

**Reply by ThomasRahm** [reply]
Ok, this is not that simple. Even just taking the outer outline could cause floating areas:
![image](https://user-images.githubusercontent.com/67757218/231651832-5efee999-8d9d-4188-be90-e6bb201b0b3b.png)
My current approach is to try to detect if a hole is valid (as in the support lines cause by it would be supported by an outline below), but this one is not even a hole.

**Reply by rburema** [reply]
That last one's caused by the `support_offset` (Support Horizontal Expansion) isn't it?

EDIT: The setting was only on for normal support previously.

**Reply by rburema** [reply]
This also causes issues like this part not actually being supported, since it would fall through the bigger ring of the tree trunk (when support interface is off). See image.
![fall_through_ring](https://user-images.githubusercontent.com/41987080/231735199-690479fc-8747-4fd5-aefd-8815b426a626.png)

**Reply by rburema** [reply]
Perhaps a solution would be to just put the restriction on support expansion back (so that it stays at `0.0`).

**Reply by ThomasRahm** [reply]
> Ok, this is not that simple. Even just taking the outer outline could cause floating areas: ![image](https://user-images.githubusercontent.com/67757218/231651832-5efee999-8d9d-4188-be90-e6bb201b0b3b.png) My current approach is to try to detect if a hole is valid (as in the support lines cause by it would be supported by an outline below), but this one is not even a hole.

I think this is "caused" by the open-close operation in WallToolPaths::generate() If i also do this (to ensure what will later be a hole is a hole when i try to filter problematic holes out) i think this behavior disappears.

That the tip may in some cases not support the model above is another issue. This can also be the case if one chooses a large tip diameter

**Reply by ThomasRahm** [reply]
So currently i again think that i may have soon a solution for the floating support parts inside of branches (branches engulfing interface causing the interface to not be supported by a support outline is another problem).
If i am correct then my largest problems are performance and that often the simulated support pattern for larger areas (when interface is disabled) should have parts removed (currently i have a workaround to prevent this but i may remove it).
I have just pushed my current state, it is very much work in progress.

**Reply by ThomasRahm** [reply]
Using a performance-optimization from PrusaSlicer i was able to get the performance for the hole filtering into what i would consider a usable state.
The issue with the lines of the top support being filtered out (correctly, as they had no support below it) i fixed by generating a fake roof for large areas. 
For this i added a few lines to FFFGcodeGenerator and SupportInfillPart (also i changed that the inset_count_to_generate now is the amount of walls generated, as i wanted the fake roof to generate without walls). If any of these seem problematic, please tell me. 
The fake roof generation is still a work in progress as there are a few cases that will need to be handled differently (e.g. high branch densities).

**Reply by ThomasRahm** [reply]
I have pushed a potential solution for high branch densities. I now generate a lower density pattern to support the high density pattern. I have an idea how to solve the other issue related to the screenshot above, but this would be then part of another improvement to better support pointy overhangs. I will update here when i know if my idea works.

**Reply by rburema** [reply]
We see it occurring far less now, since we merged in your solution. For the remaining issues, we're thinking of 'filling' the tree-support with (sparse) lightning-support-infill.

Which brings me to another point I'm not sure if I've mentioned yet; The lightning infill point generation could maybe also be repurposed (made to work with both) to generate the initial sites for the branches maybe. See the usage of the `LightningDistanceField` in `generateNewTrees` for an example.

**Reply by ThomasRahm** [reply]
I am currently trying to improve the way pointy overhang are supported, part of this is generating a high density line patterns inside of branches that support such pointy overhangs or interface (still having issues with larger areas that only touch interface being partially filled with said lines especially if the lines are to be drawn in more than one angle). This code is not yet pushed.

![image](https://user-images.githubusercontent.com/67757218/235873804-14493db8-a661-4afd-8791-5837c2a7c5f1.png)

I will need to take a look at lightning infill to see if it solves this problem in a better way.

**Reply by ThomasRahm** [reply]
I opened a draft pull request with screenshots for the improvements for pointy overhangs: https://github.com/Ultimaker/CuraEngine/pull/1880
I still do want to change some things, but it should be in a pretty good state to take a first look at.

It seems possible to use the `LightningDistanceField` to generate where tips are placed. What case do you think could be improved with it? I could see using it for small or thin overhangs to ensure more evenly tip placement, but i would need to look closer into it to be sure.

**Reply by rburema** [reply]
Thanks! I'll have a look at the PR.

As for the Lightning: We mostly think it would be a good idea to decouple the tree generation from (generic) infill, as it's not meant for that. This (also) has to do with maintenance; any change in the infill will have consequences for tree support -- conversely, any change we'll need to make to infill because of tree support will also need to work with infill. Additionally, infill isn't really made for the concept of 'try to find points to support this part', while the distance field already is.

However, on the other hand, I don't think this has high priority. If the current method works, we can also just leave it that way.

**Reply by ThomasRahm** [reply]
Supporting roof gets much more difficult without using the infill class. This includes the fake roof (similar to roof, but regular support) i use for large areas. 

I think that the most issues caused by changes in the Infill class will be if parameters are added/removed or changed compared to how support/interface lines will be generated with it.
I agree having to changed them in two places is not optimal, but i don't see any easy way to prevent this (in the tree support code) without loosing the benefits.

**Comment by rburema** [comment]
CURA-10066 Weird Effect With Adaptive Layers

**Reply by ThomasRahm** [reply]
I cant find an issue related to this one, so i do not exactly know what this weird effect is. Can i get more information about it ?

**Reply by rburema** [reply]
Of course! I don't expect you to get very far with a ticket number you can't access and a vague title.

This is a screenshot:

![image-20221220-132841](https://user-images.githubusercontent.com/41987080/223860161-62656800-a638-421d-9977-29e878057ad8.png)

This corresponds with the attached model.

[adaptive (1).zip](https://github.com/Ultimaker/Cura/files/10925595/adaptive.1.zip)

To be honest, I'm not sure how high-priority this one is. For the initial release, we can consider disabling 'Adaptive Layer Height' when printing with trees.

**Reply by ThomasRahm** [reply]
Thank you.
I see the issue here. The code assumes that each layer has a certain height (here 0.1mm) to calculate values dependent on it (e.g. diameter_angle_scale_factor,  maximum_move_distance). As here the actual layer-height can be 10% of the layer height setting, the effects are much more pronounced, causing branches to become larger and move at up to 20x rate. If reasonable settings are used (0.01mm-0.2mm layerheight is a pretty large range, with the upper limit being 20x the lower limit) the effect is much less dramatic.

I know that PrusaSlicer also has issues with the organic support and adaptive layer height, so there can be an argument made that i should look at it. 

I see two possible solutions: 
1. Making the code aware of layer-height, changing all variables that use layer-height to functions dependent on layer_idx.
2. Throwing away/unioning layers to calculate the tree support to achieve a constant distance between layers, and later extrapolate the position on the skipped layers.


The second option has the benefit that it may be also used to improve performance for small layer heights, but has a larger chance of causing issues with z distance.

I will evaluate how much work it would take to fix this as soon as i have updated my development setup.

**Reply by ThomasRahm** [reply]
I looked at possible solution 2: There are few cases that may require a good amount of changes to not cause issues. For now I will focus on the interface issue.

**Reply by rburema** [reply]
Please don't worry too much about it. I don't think supports and adaptive layer eight are used together that often.

**Comment by rburema** [comment]
CURA-10095 Slicing Crash At 100% (related to Tree Support?)

**Reply by ThomasRahm** [reply]
The corresponding issue seems to be https://github.com/Ultimaker/Cura/issues/14113

The crash seems to happen when generating Support Infill lines. It occurs in `SkeletalTrapezodiation.cpp`in `constructFromPolygons` when the `vertex` of the `twin` of `vd_edge` is `NULL`. Not sure what this means exactly, but it occurs when generating Tree Support, because it requests infill from the Infill class (Which should be fine).

In short, i think this bug is not in the TreeSupport code. I suspect it is the same bug that causes slicing failures, that can be avoided by moving the model around a bit.

**Reply by rburema** [reply]
Yes, a long standing issue known to us. We (still) think it has to do with a bug in boost-voronoi (or at least that doesn't help).

This is indeed isn't a Tree Support problem, if that's the case (though not sure why it would fail at 100% then, instead of in the middle.)

We do have two other crashes reported as well, in separate tickets, but I expect those to be much the same. (Probably at least.)

I'll attach the models here just in case it isn't.

CURA-10063: [darth slicing failed.zip](https://github.com/Ultimaker/Cura/files/10925623/darth.slicing.failed.zip)

CURA-10095: [CCR10S_santa_claus_solid_v2 (2).zip](https://github.com/Ultimaker/Cura/files/10925624/CCR10S_santa_claus_solid_v2.2.zip)

**Reply by Piezoid** [reply]
IME, this issue can be solved by scaling up the polygons before applying applying the opening/closing operation with a bigger offset. This removes more of the near-intersection ambiguities caused by snapping vertices to the grid. I've been doing that by globally scaling the unit system, but it might be easier to implement in the inputs/outputs of WallToolpath. This change allowed CuraEngine to slice every model that had this issue from the bug tracker, at least for the models that I tried.

Admittedly, this is more of a hack than a solution. The root cause may be in boost/polygon/voronoi, but CuraEngine shouldn't segfault when traversing a malformed graph. However, adding control flow for the checks to the current implementation is quite difficult.

Ideally, we should identify and fix the edge cases in the input polygons. Although, this requires understanding and reproducing the specific issue within boost/polygon/voronoi.

**Comment by rburema** [comment]
CURA-10091 Rounding Error That Causes Long Slicing Times

**Reply by ThomasRahm** [reply]
Corresponding issue seems to be https://github.com/Ultimaker/Cura/issues/14133

Fixed with https://github.com/ThomasRahm/CuraEngine/commit/201a12dbad708730aed20a0b9093c8a219079657

**Reply by rburema** [reply]
Yes, this was the ticket nr. made to pull that one in :-)

**Comment by rburema** [comment]
CURA-10161 Tree Branches Are Cut By Support Blockers

**Reply by ThomasRahm** [reply]
Assuming it is the issue i think it is, it is fixed with https://github.com/ThomasRahm/CuraEngine/commit/1d6203837a23d8c7fba1925622f27e88e305672e

**Reply by ThomasRahm** [reply]
The corresponding issue seems to be https://github.com/Ultimaker/Cura/issues/14208, but it is about wanting an option to allow tree support to pass though the support blocker, not about branches being cut/resting on support blocker.
This is not fixed as avoiding support blocker is an intentional behavior of the new tree support.
Is there a way to add a small checkbox in the support blocker UI like this and add a corresponding setting to the mesh?

Mock-up:
![image](https://user-images.githubusercontent.com/67757218/223582603-776f32cd-a966-4727-9316-390c37ec9575.png)

Otherwise the constant `SUPPORT_TREE_AVOID_SUPPORT_BLOCKER` is currently used to enable this feature, so if a global (not per mesh configurable) setting is preferable i could add it myself.

I don't think removing this behavior is beneficial, as being able to place support blocker in a hole to prevent branches from traveling though said hole can be very handy. I can provide examples/screenshots for situations where one wants the branches to avoid support blocker if you want.

**Reply by rburema** [reply]
The ticket is mostly about deciding what to actually do:

If there is no option to place a tree because of a support blocker, it just exists in the air. 
There are mixed opinions about this. 
We need to decide if we want to cull/remove these trees or give users the option to ignore the support blockers in these cases. 🤔🤔

... so we could leave it as is as well, we just need to make a clear decision.

**Reply by ThomasRahm** [reply]
The way i solved it in https://github.com/ThomasRahm/CuraEngine/commit/1d6203837a23d8c7fba1925622f27e88e305672e is that instead of using the collision when a branch has to non-graceful rest on the model, i calculate an avoidance for support blocker. That way branches that would have to rest on support blocker will not generate in the first place (the same way branches that would have to rest on the model, do not generate when the branches have to rest on the buildplate). 
I think an UI option to allow branches to travel through support blocker would be most versatile.

**Comment by rburema** [comment]
CURA-10179 Tree Support Supporting Nothing In Particular Circumstance

**Reply by ThomasRahm** [reply]
Assuming it is related to an issue like https://github.com/Ultimaker/Cura/issues/14302

Fixed with https://github.com/ThomasRahm/CuraEngine/commit/1724f67b76d1e1860df2ecaefb42a791d02e0436

**Reply by rburema** [reply]
Yes, seems to be: This ticket was made from the forums (and you replied [in that thread](https://community.ultimaker.com/topic/42362-ultimaker-cura-alpha-%F0%9F%8E%84-tree-support-spotlight-%F0%9F%8E%84/#comment-317630)).

**Comment by rburema** [comment]
Code Style & Other Refactors

CURA-10302 Make 'Raw' Magic Numbers Constant & Consolidate
CURA-10303 Streamline various 'FUDGE_LENGTH' Constants
CURA-10304 Existing Constants: Keep, Consolidate, Move To Setting?

**Comment by DerGenaue** [comment]
**Settings Default Parameters**

I've got a few questions and suggestions regarding the default parameters that were chosen for the new Tree Support:

* `Tree Support Rest Preference`: The default value of this setting is currently set to `On any flat surface`.  
  In my opinion the default should definitely be `Buildplate` to make the trees try to reach the buildplate if they can and only rest on the model if they must. I think that for most people the reduced scarring is one of the major reasons to chose Tree Supports and the slight increase in material usage compared to `On any flat surface` is perfectly justifiable.
* `Tree Support Optimal Branch Range`: This setting should in my opinion be given a higher default value. I personally even think that it should be highlighted in orange when it is smaller than `Tree Support Trunk Diameter`, because otherwise the tree trunks can overlap and this makes the Tree Supports much more prone to the floating island issue.

Are there any other opinions?
What are the motivations for the current default values?

**Reply by DerGenaue** [reply]
Regarding the `Tree Support Rest Preference`, I personally think that the best way would even be to integrate this into the `Support Placement` setting:
Three modes: `Buildplate only`, `Prefer buildplate`, `Everywhere`.
This might collide with other modes (non-tree), though...

**Reply by ThomasRahm** [reply]
I also think `Tree Support Rest Preference` should be set to `Buildplate`.

`Tree Support Optimal Branch Range` is very dependent on what is to be printed. E.g. for small miniatures i would disable it completely, but not for large models.

**Comment by ThomasRahm** [comment]
The settings related to `Support Stair Step` and `Support Join Distance` are visible when tree support is enabled, but do nothing.

**Reply by rburema** [reply]
We should decide to either make them work (probably a bit too much for for 5.4), or disable them outright (I don't think they work for our current tree-support either).

I think we can disable them, since with your tree support, it will be even less likely to need to rest on the model(s).

**Comment by ThomasRahm** [comment]
Enabling `Connect Support Lines` causes a drastic increase in tips being generated.

Enabled:
![image](https://user-images.githubusercontent.com/67757218/223587706-a28f4404-9640-4af1-99b1-c4040b345e86.png)
Disabled:
![image](https://user-images.githubusercontent.com/67757218/223587789-27050dcd-a07c-414b-9140-e7727546e220.png)
Project file:
[TestOverhang.zip](https://github.com/Ultimaker/Cura/files/10915222/TestOverhang.zip)

**Comment by ThomasRahm** [comment]
Some profiles (E.g. https://github.com/Ultimaker/Cura/blob/main/resources/definitions/flyingbear_base.def.json) seem to have `Minimum Support Area` set to something larger than 0, which can cause confusion when tree supports are used. 
Example where it has caused confusion:
https://github.com/Ultimaker/Cura/issues/14150

Also some profiles (E.g. https://github.com/Ultimaker/Cura/blob/main/resources/definitions/creality_base.def.json) seem to have a default that the `Support Top Distance` is one layer-height more than `Support Z Distance` when tree support is used. 
I don't think that this is applicable to the new tree support.

Should i make a pull request related to profiles? I had some profiles changed in my fork to accommodate for the tree support changes. For reference: https://github.com/ThomasRahm/Cura/commit/3f77b9e267517f1a24b14d15b504c7dd8154139c and https://github.com/ThomasRahm/Cura/commit/f41ed46b327dee451024561977c94aa51f848b82 (this one just fixed missing `"` in the first commit)

**Reply by rburema** [reply]
You may if you like! We'll also go over at least fdm-printer and the ultimaker ones on our side.

( @pkuiper-ultimaker Might be good to take this into account when we update our profiles. )

**Comment by ThomasRahm** [comment]
Should i also document other bugs i fixed since the xmas-alpha that were not mentioned here?

**Reply by rburema** [reply]
I think that would be helpful, yes. We can determine if we missed anything important.

**Comment by ThomasRahm** [comment]
Some areas are sometimes not supported when `Support Distance Priority` is `x/y overrides z` if other overhangs on the same layer are placed further down because of `x/y overrides z`

Fixed with https://github.com/ThomasRahm/CuraEngine/commit/2e5515de32d041aa1a30cbc719f2e2c3832d1e86

**Comment by ThomasRahm** [comment]
`Inital layer diameter` not working reliably when `rest preference` is `on any flat surface`

Fixed with https://github.com/ThomasRahm/CuraEngine/commit/e81807b00950ce24840e726b5092c4a857d20fe4 and https://github.com/ThomasRahm/CuraEngine/commit/4d65ae7dd955a29375e329c43b9493582987b18b

Very remotely related: There is a typo in `Inital Layer Diameter`, should be `Initial Layer Diameter`. The typo is not fixed.

**Comment by rburema** [comment]
The branches we're going to use for our 'exploratory' (and merge to main if we think it's close enough) are:

- For the engine; [CURA-10415_update_tree](https://github.com/Ultimaker/CuraEngine/tree/CURA-10415_update_tree)
- For the front-end; [CURA-10415_main_based_update_tree](https://github.com/Ultimaker/Cura/tree/CURA-10415_main_based_update_tree) -- Note that this is a different branch name from the engine! The one that _is_ the same as the engine was (also) based off-of the 'xmas alpha', and there where some changes we where dubious about in there, so we made a new branch that only has main and the tree changes in it.

(@ThomasRahm Jelle asked me to specifically tell you about the updated frontend-branch.)

**Reply by ThomasRahm** [reply]
Thank you. I noticed in the UI branch that quite a few changes were not migrated. I have done a pull request https://github.com/Ultimaker/Cura/pull/15074. it also contains the profile changes i mentioned earlier. If any changes were that intentional that i this pull request would undo (e.g. Disabling minimum support area) please tell me.

**Reply by rburema** [reply]
Thanks for the additional fixes (not just the UI one I mean). We'll take another look at the settings with the Print Processing & Materials team when we have resolved some of the things we still want to fix.

**Comment by ThomasRahm** [comment]
Was informed about a bug related to roof generation:
![image](https://user-images.githubusercontent.com/67757218/228437236-5dbe75f9-516b-4da7-8dc3-32aea16a5953.png)

I will be looking into this

Source: https://github.com/ThomasRahm/CuraEngine/issues/11
Some settings have to be adjusted (e.g. support roof wall line count to 0) when trying to reproduce this one.

**Reply by ThomasRahm** [reply]
Fixed with https://github.com/ThomasRahm/CuraEngine/commit/d046eafb2a87662d2165584f026df3b6827dea12

**Comment by ThomasRahm** [comment]
Roof lines that are directly next to support blocker are not supported.
Will be looking into this one. I expect a rounding error.

![image](https://user-images.githubusercontent.com/67757218/228611282-7c51ec6a-24d8-48b9-b9ed-2b39d8dc81aa.png)

**Reply by ThomasRahm** [reply]
It was not a rounding error. It is fixed with https://github.com/ThomasRahm/CuraEngine/commit/f20f751a39eab7ba07e1f41758a4487c168cf221 though.

**Comment by rburema** [comment]
@ThomasRahm I'm not sure how closely you follow our branches, but we're currently working on CURA-10415_update_tree -- We're working on some bugs, some of which is due to the mess I created with my supposed divide-by-zero fix. (Way less support was generated under some circumstances.) This should be fixed however with the same set of changes that where meant to seep up the code, as I had to remove that fix anyway.

**Comment by ThomasRahm** [comment]
I found a bug: If config.getRadius(elem) is called while elem is a pointer the constructor used to initialize a new support element on the layer below is used to convert the pointer into an element.
So every time a pointer is used to call these functions it will be slightly too large (if it is part of a tip).
I made the TreeSupportElement constructor that takes a TreeSupportElement* explicit, and solved all compile errors de-referencing the pointers instead of passing them directly. 
As i noticed this pretty far into me changing something else, it is not in a cherry-pickable commit ( https://github.com/ThomasRahm/CuraEngine/commit/dc64f9338e8cfaa9d291e54d43ed2894aa0b6fdc ).

Should i "backport" the fix for the current main branch and open another pull request?
